graspy/align/seedless_procrustes.py

```diff
@@ -0,0 +1,353 @@
+                raise TypeError(msg)
+            initial_Q = check_array(initial_Q, copy=True)
+            if initial_Q.shape[0] != initial_Q.shape[1]:
+                msg = "Initial_Q must be a square orthogonal matrix"
+                raise ValueError(msg)
+            if not np.allclose(initial_Q.T @ initial_Q, np.eye(initial_Q.shape[0])):
+                msg = "Initial_Q must be a square orthogonal matrix"
+                raise ValueError(msg)
+        # check initial_P argument
+        if initial_P is not None:
+            if not isinstance(initial_P, np.ndarray):
+                msg = f"Initial_P must be np.ndarray or None, not {type(initial_P)}"
+                raise TypeError(msg)
+            initial_P = check_array(initial_P, copy=True)
+            n, m = initial_P.shape
+            if not (
+                np.allclose(initial_P.sum(axis=0), np.ones(m) / m)
+                and np.allclose(initial_P.sum(axis=1), np.ones(n) / n)
+            ):
+                msg = (
+                    "Initial_P must be a doubly stochastic matrix "
+                    "(rows add up to (1/number of cols) "
+                    "and columns add up to (1/number of rows))"
+                )
+                raise ValueError(msg)
+
+        super().__init__()
+
+        self.optimal_transport_eps = optimal_transport_eps
+        self.optimal_transport_num_reps = optimal_transport_num_reps
+        self.optimal_transport_lambda = optimal_transport_lambda
+        self.iterative_num_reps = iterative_num_reps
+        self.init = init
+        self.initial_Q = initial_Q
+        self.initial_P = initial_P
+
+    def _optimal_transport(self, X, Y, Q):
+        # "E step" of the SeedlessProcrustes.
+        n, d = X.shape
+        m, _ = Y.shape
+        # initialize probability mass arrays & the cost matrix ; run sinkhorn
+        probability_mass_X = np.ones(n) / n
+        probability_mass_Y = np.ones(m) / m
+        cost_matrix = (
+            np.linalg.norm((X @ Q).reshape(n, 1, d) - Y.reshape(1, m, d), axis=2) ** 2
+        )
+        P = ot.sinkhorn(
+            a=probability_mass_X,
+            b=probability_mass_Y,
+            M=cost_matrix,
+            reg=self.optimal_transport_lambda,
+            numItermax=self.optimal_transport_eps,
+            stopThr=self.optimal_transport_eps,
+        )
+        return P
+
+    def _procrustes(self, X, Y, P):
+        # "M step" of the SeedlessProcurstes.
+        aligner = OrthogonalProcrustes()
+        Q = aligner.fit(X, P @ Y).Q_
+        return Q
+
+    def _iterative_ot(self, X, Y, Q):
+        for i in range(self.iterative_num_reps):
+            P = self._optimal_transport(X, Y, Q)
+            Q = self._procrustes(X, Y, P)
+            c = self._compute_objective(X, Y, Q, P)
+        return P, Q
+
+    def _compute_objective(self, X, Y, Q=None, P=None):
+        if Q is None:
+            Q = self.Q_
+        if P is None:
+            P = self.P_
```

Comment:
I find this a bit weird that sometimes this would use a class attribute and sometimes a passed value, @dwaynepryce any comments?

Reply:
happy to always pass instead.

Reply:
if u insist

Reply:
i thought it's nice to have it flexible: it's not exposed to users anyway, and i think the behavior is quite clear to any programmer at first sight. but maybe im mistaken